hibernate-core/src/main/java/org/hibernate/query/sqm/tree/expression/AbstractSqmExpression.java

```diff
@@ -94,6 +93,9 @@ public SqmExpression<String> asString() {
 
 	@Override
 	public <X> SqmExpression<X> as(Class<X> type) {
+		if ( getExpressible().getRelationalJavaType().getJavaType() == type ) {
```

Comment:
Can `getExpressible()` return `null`? Also, to be safe, we should probably use `getJavaTypeClass()` to avoid parameterized raw type issues. Another important thing is to consider is that a cast to e.g. `Duration` should not do a physical cast, just because the relational java type `Long != Duration`, so we need to retrieve the relational java type for the basic type of the passed java type class.

```suggestion
		if ( getExpressible() != null && getExpressible().getRelationalJavaType().getJavaTypeClass() == nodeBuilder().getTypeConfiguration().getBasicTypeForJavaType( type ).getRelationalJavaType().getJavaTypeClass() ) {
```

Since this method ought to return an expression of type `X` i.e. result values should be of the correct type when queried, I still think we need some sort of wrapper expression like in https://github.com/hibernate/hibernate-orm/pull/5634, otherwise a query like `cq.select(root.get("name").as(Integer.class))` will fail with a class cast exception at runtime.